lib/src/rules/dart_imports_go_first.dart

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.dart_imports_go_first;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+const desc = r' ';
+
+const details = r'''
+
+**DO** place “*dart:*” imports before other imports.
+
+**BAD:**
+```
+import 'package:bar/bar.dart';
+import 'package:foo/foo.dart';
+
+import 'dart:async';  // LINT
+import 'dart:html';  // LINT
+```
+
+**BAD:**
+```
+import 'dart:html';  // OK
+import 'package:bar/bar.dart';
+
+import 'dart:async';  // LINT
+import 'package:foo/foo.dart';
+```
+
+**GOOD:**
+```
+import 'dart:async';  // OK
+import 'dart:html';  // OK
+
+import 'package:bar/bar.dart';
+import 'package:foo/foo.dart';
+```
+
+''';
+
+class DartImportsGoFirst extends LintRule {
+  _Visitor _visitor;
+
+  DartImportsGoFirst()
+      : super(
+            name: 'dart_imports_go_first',
+            description: desc,
+            details: details,
+            group: Group.style) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    Iterable<ImportDirective> nodes =
+        DartTypeUtilities.traverseNodesInDFS(node).where(_isImportDirective);
```

Comment:
It would be much more efficient here for the initializer to be `nodes = node.directives;`.